internal/controller/bucket_controller_test.go

```diff
@@ -0,0 +1,223 @@
+		fakeClient = fake.NewClientBuilder().Build()
+		cloudServices = cloudfake.FakeCloudServices{
+			FakeS3Service: &cloudfake.FakeS3Service{},
+		}
+
+		// creates dummy bucket
+		bucket = v1alpha1.Bucket{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      BucketName,
+				Namespace: BucketNamespace,
+			},
+			Spec: v1alpha1.BucketSpec{
+				Name: BucketName,
+			},
+		}
+		_ = fakeClient.Create(ctx, &bucket)
+	})
+
+	var _ = Describe("CAPA", func() {
+		// creates the reconciler
+		BeforeEach(func() {
+			reconciler = &BucketReconciler{
+				Client:        fakeClient,
+				CloudServices: &cloudServices,
+				ManagementCluster: managementcluster.ManagementCluster{
+					Name:      "test-mc",
+					Namespace: "giantswarm",
+					Provider:  "capa",
+					Region:    "eu-central-1",
+				},
+			}
+		})
+
+		// creates a dummy management cluster and management cluster identity
+		JustBeforeEach(func() {
+			clusterIdentity := &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"kind":       "AWSClusterRoleIdentity",
+					"apiVersion": "infrastructure.cluster.x-k8s.io/v1beta2",
+					"metadata": map[string]interface{}{
+						"name":      reconciler.ManagementCluster.Name,
+						"namespace": reconciler.ManagementCluster.Namespace,
+					},
+					"spec": map[string]interface{}{
+						"roleARN": "role",
+					},
+				},
+			}
+			_ = fakeClient.Create(ctx, clusterIdentity)
+
+			cluster := &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"kind":       "AWSCluster",
+					"apiVersion": "infrastructure.cluster.x-k8s.io/v1beta2",
+					"metadata": map[string]interface{}{
+						"name":      reconciler.ManagementCluster.Name,
+						"namespace": reconciler.ManagementCluster.Namespace,
+					},
+					"spec": map[string]interface{}{
+						"identityRef": map[string]interface{}{
+							"name": reconciler.ManagementCluster.Name,
+						},
+					},
+				},
+			}
+			_ = fakeClient.Create(ctx, cluster)
+
+			// starts the reconciler
+			request := ctrl.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      BucketName,
+					Namespace: BucketNamespace,
+				},
+			}
+			_, reconcileErr = reconciler.Reconcile(ctx, request)
+		})
+
+		When("the cluster has an identity set", func() {
+			When("reconciling a missing bucket", func() {
+				BeforeEach(func() {
+					_ = fakeClient.Delete(ctx, &bucket)
+				})
+				It("does nothing", func() {
+					Expect(reconcileErr).NotTo(HaveOccurred())
+					var existingBucket v1alpha1.Bucket
+					_ = fakeClient.Get(ctx, client.ObjectKeyFromObject(&bucket), &existingBucket)
+					Expect(existingBucket.Finalizers).To(BeEmpty())
+				})
+			})
+
+			When("reconciling a new s3 bucket", func() {
+				It("was created", func() {
+					Expect(reconcileErr).NotTo(HaveOccurred())
+					var existingBucket v1alpha1.Bucket
+					_ = fakeClient.Get(ctx, client.ObjectKeyFromObject(&bucket), &existingBucket)
+					Expect(existingBucket.Finalizers).ToNot(BeEmpty())
+				})
```

Comment:
Here we are only testing that the reconciler hasn't returned an error and that there are finalizers. Maybe we can make sure we are calling the storage service? And checking the right finalizer is in the list?

Reply:
We can, sure

Reply:
How do you ensure a service was called ? I'm really starting out with Ginkgo :D